x/merkledb/db.go

```diff
@@ -65,8 +65,64 @@ type Config struct {
 	Tracer trace.Tracer
 }
 
-// Database can only be edited by committing changes from a trieView.
-type Database struct {
+type MerkleDB interface {
+	database.Database
+	Trie
+
+	// GetChangeProof returns a proof for a subset of the key/value changes in key range
+	// [start, end] that occurred between [startRootID] and [endRootID].
+	// Returns at most [maxLength] key/value pairs.
+	GetChangeProof(
+		ctx context.Context,
+		startRootID ids.ID,
+		endRootID ids.ID,
+		start []byte,
+		end []byte,
+		maxLength int,
+	) (*ChangeProof, error)
+
+	// Returns nil iff all of the following hold:
+	//   - [start] <= [end].
+	//   - [proof] is non-empty iff [proof.HadRootsInHistory].
+	//   - All keys in [proof.KeyValues] and [proof.DeletedKeys] are in [start, end].
+	//     If [start] is empty, all keys are considered > [start].
+	//     If [end] is empty, all keys are considered < [end].
+	//   - [proof.KeyValues] and [proof.DeletedKeys] are sorted in order of increasing key.
+	//   - [proof.StartProof] and [proof.EndProof] are well-formed.
+	//   - When the keys in [proof.KeyValues] are added to [db] and the keys in [proof.DeletedKeys]
+	//     are removed from [db], the root ID of [db] is [expectedEndRootID].
+	//
+	// Assumes [db.lock] isn't held.
+	// This is defined on Database instead of ChangeProof because it accesses
+	// database internals.
+	VerifyChangeProof(
+		ctx context.Context,
+		proof *ChangeProof,
+		start []byte,
+		end []byte,
+		expectedEndRootID ids.ID,
+	) error
+
+	// CommitChangeProof commits the key/value pairs within the [proof] to the db.
+	CommitChangeProof(ctx context.Context, proof *ChangeProof) error
```

Comment:
This didn't change in this PR, but yeah, this errors if the change proof is invalid

Reply:
My mistake above -- this actually doesn't verify the change proof. That's done elsewhere. This will only error if there's some issue actually writing the change proof. For all of our database implementations, we consider a non-nil error on write to be fatal. That is, it's never expected to happen, and if it does, the application using it should stop. In fact, we wrap databases to enforce that if an unexpected error occurs (i.e. not `database.ErrNotFound` or `database.ErrClosed`) then the database will refuse any subsequent operations. (See `database/corruptabledb`.) 

@StephenButtolph Do you think it's worth adding an invariant on `database.Database` that all non-nil errors on writes should be treated as fatal?